[HDR] Enable HDR display for CSS images

shallawa opened the pull request:
#### c680aebea33bf4560a419eb2c30798a85c0b56f6
<pre>
[HDR] Enable HDR display for CSS images
<a href="https://bugs.webkit.org/show_bug.cgi?id=289363">https://bugs.webkit.org/show_bug.cgi?id=289363</a>
<a href="https://rdar.apple.com/146509866">rdar://146509866</a>

Reviewed by Simon Fraser.

The CSS images needs to be checked to decide whether rendererHasHDRContent()
returns true. Headroom::FromImage will be passed in the ImagePaintingOptions
when drawing the HDR CSS images.

RenderElement::imageContentChanged() will setHasHDRContent() on Document. It
will also mark contentChanged(Image) on its enclosingCompsitedLayer().

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-background-image-expected.txt: Renamed from LayoutTests/platform/ios/compositing/hdr/hdr-css-image-expected.txt.
* LayoutTests/compositing/hdr/hdr-background-image.html: Copied from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/compositing/hdr/hdr-border-image-expected.txt: Renamed from LayoutTests/platform/mac-wk2/compositing/hdr/hdr-css-image-expected.txt.
* LayoutTests/compositing/hdr/hdr-border-image.html: Copied from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/compositing/hdr/hdr-pseudo-before-image.html: Renamed from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/fast/images/hdr-background-image-expected.html: Renamed from LayoutTests/fast/images/hdr-css-image-expected.html.
* LayoutTests/fast/images/hdr-background-image.html: Copied from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/fast/images/hdr-basic-image-expected.html:
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-border-image-expected.html: Copied from LayoutTests/fast/images/hdr-basic-image-expected.html.
* LayoutTests/fast/images/hdr-border-image.html: Copied from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/fast/images/hdr-pseudo-before-image-expected.html: Copied from LayoutTests/fast/images/hdr-basic-image-expected.html.
* LayoutTests/fast/images/hdr-pseudo-before-image.html: Renamed from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt:
* LayoutTests/platform/ios/compositing/hdr/hdr-pseudo-before-image-expected.txt: Copied from LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-expected.txt:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-pseudo-before-image-expected.txt: Copied from LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::drawPattern):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::notifyFinished):
(WebCore::RenderElement::imageContentChanged):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::notifyFinished):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::hasHDRContent const):
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::notifyFinished):

Canonical link: <a href="https://commits.webkit.org/292305@main">https://commits.webkit.org/292305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4105553260204777dde025b01e9c117050ca668e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100608 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97603 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23596 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72878 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30144 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11575 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11281 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4016 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45400 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81472 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102643 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22609 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16515 "Found 1 new test failure: imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81920 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22861 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82318 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81271 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20362 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25850 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3321 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15941 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22577 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27734 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22236 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25712 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5790,27 +5790,31 @@ static bool isRenderElementWithHDR(const RenderElement& renderer)
     }
 #endif
 
+    auto& style = renderer.style();
```
`CheckedRef`

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -507,6 +507,15 @@ RenderLayer* RenderObject::enclosingLayer() const
     return nullptr;
 }
 
+RenderLayer* RenderObject::enclosingCompsitedLayer() const
```
Typo: should be `enclosingCompositedLayer` but I think you should remove this.

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1719,12 +1722,14 @@ void RenderElement::imageContentChanged(CachedImage& cachedImage)
     }
+
+    if (document().hasPaintedHDRContent()) {
+        if (auto layer = enclosingCompsitedLayer())
```
I think it's better to keep using `enclosingLayer()`, and have the RenderLayer code find the enclosing compositing layer if necessary. RenderElement shouldn't need to think about compositing.

shallawa replied:
Can you please explain a little bit what is needed here? Do you mean adding a flag to `enclosingLayer(bool composited)`?

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1719,12 +1722,14 @@ void RenderElement::imageContentChanged(CachedImage& cachedImage)
             document().setHasPaintedHDRContent();
     }
+
+    if (document().hasPaintedHDRContent()) {
```
Every time I read "hasPaintedHDRContent" I have to think "Does it mean 'painted HDR content sometime in the past', or 'owns HDR content which is painted'".

shallawa replied:
Yes you are right. The names should be cleaned such that we do not have `hasPaintedHDRContent` and `drawsHDRContent` etc. I do that in a separate PR.

shallawa replied:
The naming is addressed in https://github.com/WebKit/WebKit/pull/42556.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5829,26 +5829,36 @@ static bool isRenderElementWithHDR(const RenderElement& renderer)
+        }
+    }
+
+    if (style->hasBorderImage()) {
```
What about `listStyleImage()`? There may be other images in style I'm forgetting. `RenderElement::styleDidChange()` does some stuff for all the images; you need to do the same. Note also the `getCachedPseudoStyle()` there which you should copy.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5829,26 +5829,36 @@ static bool isRenderElementWithHDR(const RenderElement& renderer)
+    CheckedRef style = renderer.style();
+
+    if (style->hasBackgroundImage()) {
+        if (auto* cachedImage = style->backgroundLayers().image()->cachedImage()) {
```
`backgroundLayers()` is a list of FillLayers; you need to traverse the list. Your test should test multiple background images.

shallawa replied:
Fixed.

smfr commented on LayoutTests/compositing/hdr/hdr-background-image.html:
```diff
@@ -56,16 +49,15 @@
 
             containerElements[0].style.backgroundImage = 'url(' + images[0].src + ')';
-            containerElements[1].style.borderImage = 'url(' + images[0].src + ') 30 fill';
+            containerElements[1].style.backgroundImage = 'url(' + images[1].src + ')';
```
Why not just do this with declarative CSS via class names?

smfr commented on LayoutTests/fast/images/hdr-background-image.html:
```diff
@@ -55,16 +48,15 @@
             containerElements[0].style.backgroundImage = 'url(' + images[0].src + ')';
-            containerElements[1].style.borderImage = 'url(' + images[0].src + ') 30 fill';
+            containerElements[1].style.backgroundImage = 'url(' + images[1].src + ')';
+            containerElements[2].style.backgroundImage = 'url(' + images[1].src + ')' + ', url(' + images[0].src + ')';
```
Ditto

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -507,6 +507,15 @@ RenderLayer* RenderObject::enclosingLayer() const
     return nullptr;
 }
 
+RenderLayer* RenderObject::enclosingCompositedLayer() const
```
I'd still prefer that you don't add this. Talk to RenderLayer, and have it find the enclosing composited layer.

nt1m commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5829,26 +5829,54 @@ static bool rendererHasHDRContent(const RenderElement& renderer)
+    for (auto pseudoId = PseudoId::FirstPublicPseudoId; pseudoId < PseudoId::FirstInternalPseudoId; pseudoId = static_cast<PseudoId>(static_cast<unsigned>(pseudoId) + 1)) {
+        if (styleHasHDRContent(renderer.getCachedPseudoStyle({ pseudoId })))
+            return true;
     }
```
Why are we doing this? Shouldn't pseudo-elements be like normal element children?

Also the iteration here won't work properly when targeting pseudo-elements with arguments like highlights or view transitions.

shallawa committed:
[HDR] Enable HDR display for CSS images
https://bugs.webkit.org/show_bug.cgi?id=289363
rdar://146509866

Reviewed by Simon Fraser.

The CSS images needs to be checked to decide whether rendererHasHDRContent()
returns true. Headroom::FromImage will be passed in the ImagePaintingOptions
when drawing the HDR CSS images.

RenderElement::imageContentChanged() will setHasHDRContent() on Document. It
will also mark contentChanged(Image) on its enclosingCompsitedLayer().

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-background-image-expected.txt: Renamed from LayoutTests/platform/ios/compositing/hdr/hdr-css-image-expected.txt.
* LayoutTests/compositing/hdr/hdr-background-image.html: Copied from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/compositing/hdr/hdr-border-image-expected.txt: Renamed from LayoutTests/platform/mac-wk2/compositing/hdr/hdr-css-image-expected.txt.
* LayoutTests/compositing/hdr/hdr-border-image.html: Copied from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/compositing/hdr/hdr-pseudo-before-image.html: Renamed from LayoutTests/compositing/hdr/hdr-css-image.html.
* LayoutTests/fast/images/hdr-background-image-expected.html: Renamed from LayoutTests/fast/images/hdr-css-image-expected.html.
* LayoutTests/fast/images/hdr-background-image.html: Copied from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/fast/images/hdr-basic-image-expected.html:
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-border-image-expected.html: Copied from LayoutTests/fast/images/hdr-basic-image-expected.html.
* LayoutTests/fast/images/hdr-border-image.html: Copied from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/fast/images/hdr-pseudo-before-image-expected.html: Copied from LayoutTests/fast/images/hdr-basic-image-expected.html.
* LayoutTests/fast/images/hdr-pseudo-before-image.html: Renamed from LayoutTests/fast/images/hdr-css-image.html.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt:
* LayoutTests/platform/ios/compositing/hdr/hdr-pseudo-before-image-expected.txt: Copied from LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-expected.txt:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-pseudo-before-image-expected.txt: Copied from LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-expected.txt.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::drawPattern):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::notifyFinished):
(WebCore::RenderElement::imageContentChanged):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::notifyFinished):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::hasHDRContent const):
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::notifyFinished):

Canonical link: https://commits.webkit.org/292305@main
